# Main entrypoint of the workflow.
# Please follow the best practices:
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there.


# load configuration
# -----------------------------------------------------
configfile: "config/config.yaml"


# load rules
# -----------------------------------------------------


module fastq_process_align:
    snakefile:
        github("aryazand/fastq-process-align", path="workflow/Snakefile", branch="main")
    config:
        config


use rule * from fastq_process_align as fastq_process_align_*


trackhub_dir = "genome_browser/ucsc_trackhub/"


include: "rules/star2ndpass.smk"
include: "rules/stringtie.smk"
include: "rules/mapping_stats.smk"
include: "rules/ucsc_trackhub.smk"
include: "rules/coverage.smk"


# optional messages, log and error handling
# -----------------------------------------------------
onstart:
    print("\n--- Analysis started ---\n")


onsuccess:
    print("\n--- Workflow finished! ---\n")


onerror:
    print("\n--- An error occurred! ---\n")


# target rules
# -----------------------------------------------------
rule all:
    input:
        "results/multiqc/multiqc_report.html",
        "results/stringtie/stringtie_merged.gtf",
        "results/gffcompare/gffcmp.annotated.gtf",
        expand(
            os.path.join(trackhub_dir, "{org}", "trackDb.txt"),
            org=config["ucsc_trackhub"]["genomes"],
        ),
    default_target: True
